gateway/ln-gateway/src/lib.rs

```diff
@@ -511,112 +511,117 @@ impl Gateway {
     }
 
     pub async fn run(mut self, listen: SocketAddr, password: String) -> Result<()> {
-        let mut tg = self.task_group.clone();
-
         let sender = GatewayRpcSender::new(self.sender.clone());
-        tg.spawn("Gateway Webserver", move |server_ctrl| async move {
-            let mut webserver = tokio::spawn(run_webserver(password, listen, sender));
-
-            // Shut down webserver if requested
-            if server_ctrl.is_shutting_down() {
-                webserver.abort();
-                let _ = futures::executor::block_on(&mut webserver);
-            }
-        })
-        .await;
+        self.task_group
+            .spawn("Gateway Webserver", move |_| async move {
+                run_webserver(password, listen, sender)
+                    .await
+                    .expect("Error starting webserver");
+            })
+            .await;
 
         // TODO: try to drive forward outgoing and incoming payments that were
         // interrupted
-        let loop_ctrl = tg.make_handle();
-        loop {
+        let loop_ctrl = self.task_group.make_handle();
+        let shutdown_sender = self.sender.clone();
+        loop_ctrl
+            .on_shutdown(Box::new(move || {
+                Box::pin(async move {
+                    // ignore errors: we don't care if already stopped
+                    let _ = shutdown_sender.send(GatewayRequest::Shutdown()).await;
+                })
+            }))
+            .await;
+
+        // Handle messages from webserver and plugin
+        while let Some(msg) = self.receiver.recv().await {
+            tracing::trace!("Gateway received message {:?}", msg);
+
```

Comment:
One could create a shutdown receiver instead here, which internally does pretty much the same as the manual `on_shutdown` registration, then once could `select!` over either `self.receiver.recv()` or the shutdown receiver receiving something. But this works just as well.